numba/testing/main.py

```diff
@@ -807,3 +808,62 @@ def run(self, test):
         # This will call self._run_inner() on the created result object,
         # and print out the detailed test results at the end.
         return super(ParallelTestRunner, self).run(self._run_inner)
+
+
+
+
+import xmlrunner
+class ParallelXMLTestRunner:
+    """
+    A test runner which delegates the actual running to a pool of child
+    processes.
+    """
+    timeout = _TIMEOUT
+
+    def __init__(self, *, nprocs, useslice, **kwargs):
+        self.nprocs = nprocs
+        self.useslice = parse_slice(useslice)
+        self._kwargs = kwargs.copy()
+
+    def run(self, test):
+        self._ptests, self._stests = _split_nonparallel_tests(test,
+                                                              sliced=self.useslice)
+        print("Parallel: %s. Serial: %s" % (len(self._ptests),
+                                            len(self._stests)))
+        runner = xmlrunner.XMLTestRunner(**self._kwargs)
+        return runner.run(self._run_inner)
+
+    def _run_inner(self, result):
+        pool = multiprocessing.get_context("spawn").Pool(self.nprocs)
+
+        ptests = self._ptests
+        if ptests:
+            chunk_size = (len(ptests) + self.nprocs - 1) // self.nprocs
+            splitted_tests = [ptests[i:i + chunk_size]
+                            for i in range(0, len(self._ptests), chunk_size)]
+            assert sum(map(len, splitted_tests)) == len(ptests)
+
+            for idx, status in pool.imap_unordered(xml_child_runner, enumerate(splitted_tests)):
+                if not status:
+                    case = type(f"ParallelTest{idx}", (_FailedTest,), {})
+                    ss = loader.defaultTestLoader.loadTestsFromTestCase(case)
+                    ss.run(result)
+
+        stests = SerialSuite(self._stests)
+        stests.run(result)
+        return result
```

Comment:
This part may have to look a bit like:


```

    def _run_inner(self, result):
        pool = multiprocessing.get_context("spawn").Pool(self.nprocs)

        ptests = self._ptests
        print("len(ptests)", len(ptests))

        # Split the tests and recycle the worker process to tame memory usage.
        chunk_size = 100
        splitted_tests = [self._ptests[i:i + chunk_size]
                        for i in range(0, len(self._ptests), chunk_size)]

        for tests in splitted_tests:
            pool = multiprocessing.Pool(self.nprocs)
            try:
                result = xml_child_runner(tests)
            except:
                # On exception, kill still active workers immediately
                pool.terminate()
                # Make sure exception is reported and not ignored
                raise
            else:
                # Close the pool cleanly unless asked to early out
                if result.shouldStop:
                    pool.terminate()
                    break
                else:
                    pool.close()
            finally:
                # Always join the pool (this is necessary for coverage.py)
                pool.join()
        if not result.shouldStop:
            stests = SerialSuite(self._stests)
            stests.run(result)
            return result
        return result
```

which is mostly a copy of https://github.com/numba/numba/blob/bfe1270e23f8daf7fd95cfb6ef385f6c832fd927/numba/testing/main.py#L744-L776